### ISSUE TITLE
[Feature Request] Add Click'n'Load 2 support

### ISSUE THREAD

**Issue by Anuskuss** [jlesage/docker-jdownloader-2#7]
First of all, thanks for your work. To make the experience even better I'm suggesting to add Click'n'Load support. You can already redirect 127.0.0.0:9666 from Windows to your server, the application just has to listen for it :smiley:

**Comment by rix1337** [comment]
Already implemented. Use myjdownloader

**Comment by Anuskuss** [comment]
Nice, didn't know My JDownloader was capable of that. Guess I'm switching to plusminus/jdownloader-headless though.

**Comment by dauhonk** [comment]
These settings need to be changed to activate CNL for the docker JDownloader container:
- JDownloader Settings -> Advanced Settings -> Search "Flashgot" -> uncheck "will listen on localhost only" (restart required)
- Add port forwarding for CNL to Docker start arguments: “-p 9666:9666”
- On client side install “Click'n'Load request proxy” add-in and configure Docker host IP (firefox/chrome)

**Comment by rix1337** [comment]
What I noticed: MyJDownloader would use the local IP of the JDownloader, if it were not running inside docker. I cant run the container in host mode (mainly because of VNC), so the internal IP found reported to myJDownloader by this will always be the 172.X.X.X variant.

Have you guys found any way to override this with the docker host ip? This would dramatically increase the snappyness of MyJDownloader in the local network (have verified with a native JD2 install).

**Comment by jlesage** [comment]
I guess you are talking about the MyJDownloader Browser Extension?

**Comment by rix1337** [comment]
And the mobile app.

Both will try to reach the docker bridge subnet first, the myjdownloader connection second (relayed over WAN). Since the bridge subnet has an IP range unavailable to the LAN, both the extension and app have constant connection issues. They work, but slowly.

Any idea where we can override or spoof the docker host IP "into jdownloader"?

**Comment by pokrzys** [comment]
> What I noticed: MyJDownloader would use the local IP of the JDownloader, if it were not running inside docker, when contacted in lan. I cant run the container in host mode (mainly because of VNC), so the internal IP reported to myJDownloader by this will always be the 172.X.X.X variant.

Thanks for the great find! I also found that standalone install of JD2 is running much better in MyJdownloader than the docker one. Just didn't realize that using host instead of brigde would fix this problem! I've nothing against running it at host network so now my JD2 is running almost as smooth as standard installation.

**Comment by pokrzys** [comment]
I am running it at nas with openmediavault as OS. It has plugin for docker with nice GUI so I just switched from Bridge to Host in that GUI, restarted container and that's all. But I don't use vnc so I don't know if it can couse any problems.

**Comment by pokrzys** [comment]
Have you tried another jdownloader image? On another machine I'm using plusminus/jdownloader-headless which also works fine. It just doesn't have webpage, being used just through MyJdownloader website. Can't decide which one top use so I use both. It is jd for headless systems and should not have any conflicts with x-server.

**Comment by jlesage** [comment]
Please try the following steps to get a direct connection.  It worked for me.

  - First, add a new port mapping for TCP port `3129` to your container configuration (`-p 3129:3129`).
  - Open the JDownloader UI.
  - Go to `Settings`->`Advanced Settings`.
  - Search for `MyJDownloaderSettings`.
  - Set `Custom Device IPs` to the IP address of the host running the container, between double quotes (e.g. `"192.168.1.1"`).
  - Change the `Direct Connection Mode` to `Allow lan/wan connections with manual port forwarding`.
  - Restart JDownloader.

**Comment by rix1337** [comment]
Oh god, this is so much better!!! Thank you @jlesage this is all I ever wanted!

**Comment by tristan-k** [comment]
Do I still need a MyJDownloader account to communicate between my docker container with jdownloader and my browser running on a different computer?

**Comment by dauhonk** [comment]
> Do I still need a MyJDownloader account to communicate between my docker container with jdownloader and my browser running on a different computer? I dont want to share data my data with my.jdownloader.org for privacy reasons.

No. If the ports below are forwarded to your network and your client computer is on the same network you can directly communicate with jdownloader:
5800/tcp --> WebGUI (your Browser)
5900/tcp --> VNC (e.g. tightVNC Client)
9666/tcp --> CNL (via Browser CNL plugin or SSH tunnel (e.g. PuTTY))

No need for a myjdownloader account

**Comment by tristan-k** [comment]
Thanks for answering. Can you recommend a click'n'load firefox addon and explain how to configure it? Also is it possible to redirect and solve captchas with them?

**Comment by zensbert** [comment]
Dauhonks approch works perfect:

- JDownloader Settings -> Advanced Settings -> Search "Flashgot" -> uncheck "will listen on localhost only" (restart required)
- Add port forwarding for CNL to Docker start arguments: “-p 9666:9666”

But i wanted to add:

- On your "windows client" redirect the local port 9666 to the jdownloader docker image by typing in CMD:

netsh interface portproxy add v4tov4 9666 IP_OF_DOCKER 9666 127.0.0.1
